### ISSUE TITLE
NotImplemented is incompatible with None

### ISSUE THREAD

**Issue by evtn** [microsoft/pyright#7774]
**Describe the bug**
When writing a base class function that is not implemented, but should normally return None, pyright complains about that:

```python
class Base:
    def send_bread_to_space(self) -> None:
        return NotImplemented
    
    def send_car_to_space(self) -> int:
        return NotImplemented
    
    def send_space_to_car(self) -> str:
        return NotImplemented

# todo: add SpaceSender class
```

This sample will produce one error: 

`error: Expression of type "_NotImplementedType" cannot be assigned to return type "None"`

...but all other types will work fine.

**Pyright info**
version: 1.1.360 (cli)

**Comment by erictraut** [comment]
This is intended behavior. According to the typeshed stubs, `NotImplemented` is an instance of the class `_NotImplementedType` which is a special class that derives from `Any`. In general, classes that drive from `Any` must be considered compatible with all other classes unless those classes are `final`. The `NoneType` class is final, so it cannot possibly be compatible with `_NotImplementedType`. In practice, it's problematic to apply this rule to all `final` classes, but it is important for a variety of other reasons for the rule to apply to `None` in particular. A bunch of other common use cases break without this special case.

FWIW, mypy also applies this special case, and its behavior matches pyright's for the code in your example.
